File: leetcode/450_Delete_Node_in_a_BS.py
from unittest import TestCase

from typing import Optional

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

    def __eq__(self, o):
        return isinstance(o, TreeNode) and self.val == o.val and self.left == o.left and self.right == o.right

class Solution:
    def deleteNode(self, root: Optional[TreeNode], key: int) -> Optional[TreeNode]:
        def delete_node(node: Optional[TreeNode], key: int):
            if node is not None:
                if node.val < key:
                    node.right = delete_node(node.right, key)
                elif node.val > key:
                    node.left = delete_node(node.left, key)
                else:
                    if node.left is None:
                        node = node.right
                    elif node.right is None:
                        node = node.left
                    else:
                        least_node = node.right
                        while least_node.left is not None:
                            least_node = least_node.left
                        node.val = least_node.val
                        node.right = delete_node(node.right, node.val)
            return node

        return delete_node(root, key)

class DeleteNodeTest(TestCase):
    def test_1(self):
        self.assertEqual(
            Solution().deleteNode(TreeNode(5, TreeNode(3, TreeNode(2), TreeNode(4)), TreeNode(6, None, TreeNode(7))), 3), 
            TreeNode(5, TreeNode(4, TreeNode(2), None), TreeNode(6, None, TreeNode(7)))
        )

File: leetcode/450_Delete_Node_in_a_BS.rs
#[derive(Debug, PartialEq, Eq)]
pub struct TreeNode {
    pub val: i32,
    pub left: Option<Rc<RefCell<TreeNode>>>,
    pub right: Option<Rc<RefCell<TreeNode>>>,
}

impl TreeNode {
    #[inline]
    pub fn new(val: i32) -> Self {
        TreeNode {
            val,
            left: None,
            right: None
        }
    }
}
struct Solution {}

use std::rc::Rc;
use std::cell::RefCell;

impl Solution {
    pub fn delete_node(root: Option<Rc<RefCell<TreeNode>>>, key: i32) -> Option<Rc<RefCell<TreeNode>>> {

        fn find_least(node: &Rc<RefCell<TreeNode>>) -> i32 {
            if let Some(left) = node.borrow().left.as_ref() {
                find_least(left)
            } else {
                node.borrow().val
            }
        }

        fn delete_node(node: Option<Rc<RefCell<TreeNode>>>, key: i32) -> Option<Rc<RefCell<TreeNode>>> {
            if let Some(node) = node {
                if node.borrow().val < key {
                    let right = delete_node(node.borrow_mut().right.take(), key);
                    node.borrow_mut().right = right;
                    Some(node)
                } else if node.borrow().val > key {
                    let left = delete_node(node.borrow_mut().left.take(), key);
                    node.borrow_mut().left = left;
                    Some(node)
                } else {
                    if node.borrow().left.is_none() {
                        node.borrow_mut().right.take()
                    } else if node.borrow().right.is_none() {
                        node.borrow_mut().left.take()
                    } else {
                        let val = find_least(node.borrow().right.as_ref().unwrap());
                        node.borrow_mut().val = val;
                        let right = node.borrow_mut().right.take();
                        node.borrow_mut().right = delete_node(right, val);
                        Some(node)
                    }
                }
            } else {
                None
            }
        }
        delete_node(root, key)
    }
}
